code/modules/swarmers/swarmer.dm

```diff
@@ -0,0 +1,453 @@
+
+/**
+  * Used to determine what type of swarmer a swarmer should create
+  *
+  * Returns the type of the swarmer to be created
+  */
+/mob/living/simple_animal/hostile/swarmer/proc/swarmer_type_to_create()
+	return /mob/living/simple_animal/hostile/swarmer/melee
+
+/**
+  * Called when a swarmer attempts to repair itself
+  *
+  * Proc used to allow a swarmer self-repair.  If the swarmer does not move after a period of time, then it will heal fully
+  */
+/mob/living/simple_animal/hostile/swarmer/proc/repair_self()
+	set name = "Self Repair"
+	set category = "Swarmer"
+	set desc = "Attempts to repair damage to our body. You will have to remain motionless until repairs are complete."
+	if(!isturf(loc))
+		return
+	to_chat(src, "<span class='info'>Attempting to repair damage to our body, stand by...</span>")
+	if(!do_mob(src, src, 10 SECONDS))
+		return
+	adjustHealth(-maxHealth)
+	to_chat(src, "<span class='info'>We successfully repaired ourselves.</span>")
+
+/**
+  * Called when a swarmer toggles its light
+  *
+  * Proc used to allow a swarmer to toggle its  light on and off.  If a swarmer has any drones, change their light settings to match their master's.
+  */
+/mob/living/simple_animal/hostile/swarmer/proc/toggle_light()
+	if(!light_range)
+		set_light(3)
+		if(!mind)
+			return
+		for(var/d in dronelist)
+			var/mob/living/simple_animal/hostile/swarmer/melee/drone = d
+			drone.set_light(3)
+	else
+		set_light(0)
+		if(!mind)
+			return
+		for(var/d in dronelist)
+			var/mob/living/simple_animal/hostile/swarmer/melee/drone = d
+			drone.set_light(0)
+
+/**
+  * Proc which is used for swarmer comms
+  *
+  * Proc called which sends a message to all other swarmers.
+  * Arugments:
+  * * msg - The message the swarmer is sending, gotten from ContactSwarmers()
+  */
+/mob/living/simple_animal/hostile/swarmer/proc/swarmer_chat(msg)
+	var/rendered = "<B>Swarm communication - [src]</b> [say_quote(msg)]"
+	for(var/i in GLOB.mob_list)
+		var/mob/M = i
+		if(isswarmer(M))
+			to_chat(M, rendered)
+		else if(isobserver(M))
+			var/link = FOLLOW_LINK(M, src)
+			to_chat(M, "[link] [rendered]")
+
+/**
+  * Proc which is used for inputting a swarmer message
+  *
+  * Proc which is used for a swarmer to input a message on a pop-up box, then attempt to send that message to the other swarmers
+  */
+/mob/living/simple_animal/hostile/swarmer/proc/contact_swarmers()
+	var/message = stripped_input(src, "Announce to other swarmers", "Swarmer contact")
+	// TODO get swarmers their own colour rather than just boldtext
+	if(message)
+		swarmer_chat(message)
+		
+/**
+  * Removes a drone from the swarmer's list.
+  *
+  * Removes the drone from our list.
+  * Called specifically when a drone is about to be destroyed, so we don't have any null references.
+  * Arguments:
+  * * mob/drone - The drone to be removed from the list.
+  */
+mob/living/simple_animal/hostile/swarmer/proc/remove_drone(mob/drone, force)
+	dronelist -= drone
```

Comment:
```suggestion


///Adds a drone to the swarmer list and keeps track of it in case it's deleted and requires cleanup.
/mob/living/simple_animal/hostile/swarmer/proc/remove_drone(mob/drone
	LAZYADD(dronelist, newswarmer)
	RegisterSignal(newswarmer, COMSIG_PARENT_QDELETING, .proc/remove_drone, newswarmer)


/**
  * Removes a drone from the swarmer's list.
  *
  * Removes the drone from our list.
  * Called specifically when a drone is about to be destroyed, so we don't have any null references.
  * Arguments:
  * * mob/drone - The drone to be removed from the list.
  */
/mob/living/simple_animal/hostile/swarmer/proc/remove_drone(mob/drone, force)
	UnregisterSignal(drone, COMSIG_PARENT_QDELETING)
	dronelist -= drone
```
Make this behavior generic.